In [25]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

<h1> Zadanie 1 </h1>

In [26]:
df = pd.read_csv('Titanic_Lab2.csv', index_col = 0)
df.head(20)

pclass  survived                                               name  \
0        1         1                      Allen, Miss. Elisabeth Walton   
1        1         1                     Allison, Master. Hudson Trevor   
2        1         0                       Allison, Miss. Helen Loraine   
3        1         0               Allison, Mr. Hudson Joshua Creighton   
4        1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5        1         1                                Anderson, Mr. Harry   
6        1         1                  Andrews, Miss. Kornelia Theodosia   
7        1         0                             Andrews, Mr. Thomas Jr   
8        1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9        1         0                            Artagaveytia, Mr. Ramon   
10       1         0                             Astor, Col. John Jacob   
11       1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12       1         1                      Aubart, Mme. Leontine Pauline   
13       1         1                       Barber, Miss. Ellen 'Nellie'   
14       1         1               Barkworth, Mr. Algernon Henry Wilson   
15       1         0                                Baumann, Mr. John D   
16       1         0                           Baxter, Mr. Quigg Edmond   
17       1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18       1         1                              Bazzani, Miss. Albina   
19       1         0                               Beattie, Mr. Thomson   

       sex      age  sibsp  parch    ticket      fare    cabin embarked boat  \
0   female  29.0000      0      0     24160  211.3375       B5        S    2   
1     male   0.9167      1      2    113781  151.5500  C22 C26        S   11   
2   female   2.0000      1      2    113781  151.5500  C22 C26        S  NaN   
3     male  30.0000      1      2    113781  151.5500  C22 C26        S  NaN   
4   female  25.0000      1      2    113781  151.5500  C22 C26        S  NaN   
5     male  48.0000      0      0     19952   26.5500      E12        S    3   
6   female  63.0000      1      0     13502   77.9583       D7        S   10   
7     male  39.0000      0      0    112050    0.0000      A36        S  NaN   
8   female  53.0000      2      0     11769   51.4792     C101        S    D   
9     male  71.0000      0      0  PC 17609   49.5042      NaN        C  NaN   
10    male  47.0000      1      0  PC 17757  227.5250  C62 C64        C  NaN   
11  female  18.0000      1      0  PC 17757  227.5250  C62 C64        C    4   
12  female  24.0000      0      0  PC 17477   69.3000      B35        C    9   
13  female  26.0000      0      0     19877   78.8500      NaN        S    6   
14    male  80.0000      0      0     27042   30.0000      A23        S    B   
15    male      NaN      0      0  PC 17318   25.9250      NaN        S  NaN   
16    male  24.0000      0      1  PC 17558  247.5208  B58 B60        C  NaN   
17  female  50.0000      0      1  PC 17558  247.5208  B58 B60        C    6   
18  female  32.0000      0      0     11813   76.2917      D15        C    8   
19    male  36.0000      0      0     13050   75.2417       C6        C    A   

     body                        home.dest  KolumnaNull CabinReduced  
0     NaN                     St Louis, MO            0            B  
1     NaN  Montreal, PQ / Chesterville, ON            0            C  
2     NaN  Montreal, PQ / Chesterville, ON            0            C  
3   135.0  Montreal, PQ / Chesterville, ON            0            C  
4     NaN  Montreal, PQ / Chesterville, ON            0            C  
5     NaN                     New York, NY            0            E  
6     NaN                       Hudson, NY            0            D  
7     NaN                      Belfast, NI            0            A  
8     NaN              Bayside, Queens, NY            0            C  
9    22.0              Montevideo, Ur

<h1> Zadanie 2 </h1>

Funkcja train_test_split to bardzo uzyteczne narzedzie z pakietu Scikit-learn. Pozwala na bardzo szybki podzial danych wejsciowych na "wycinki", czyli podzbiory na ktorych bedziemy trenowac model, a nastepnie testowac, czy wyniki jego predykcji sa dla nas zadowalajace. Funkcja ma wiele uzytecznych parametrow, ktore mozemy sobie modyfikowac np. test_size, train_size - czyli w latwy i szybki sposob mzoemy okreslic, jak duze chcemy miec nasze podzbiory.

<h1> Zadanie 3</h1>

In [27]:
col_name = ['sex', 'cabin', 'CabinReduced']

<h1> Zadanie 4 </h1>

In [28]:
X = df[[c for c in col_name]]
y = df[['survived']]

In [29]:
X.head(10)

sex    cabin CabinReduced
0  female       B5            B
1    male  C22 C26            C
2  female  C22 C26            C
3    male  C22 C26            C
4  female  C22 C26            C
5    male      E12            E
6  female       D7            D
7    male      A36            A
8  female     C101            C
9    male      NaN          NaN

In [30]:
y.head(10)

survived
0         1
1         1
2         0
3         0
4         0
5         1
6         1
7         0
8         1
9         0

In [31]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0)

In [32]:
shapes = [X_train.shape, X_test.shape, y_train.shape, y_test.shape]
names = ['X_train', 'X_test', 'y_train', 'y_test']

for s, n in zip(shapes, names):
    print(f'Shape {n} : {s}')

print(df.shape)

Shape X_train : (916, 3)
Shape X_test : (393, 3)
Shape y_train : (916, 1)
Shape y_test : (393, 1)
(1309, 16)


Zgodnie z przewidywaniami - 30% calych danych przeznaczylismy na zbiory testowe. Ich dlugosc to 393 wiersze z 1309 wszystkich. Zbiory treningowe dla modelu to pozostala czesc - 916 wierszy.

<h1> Zadanie 5 </h1>

In [33]:
for c in col_name:
    unique_test = [x for x in X_test[c].unique() if x not in X_train[c].unique()]
    print(f'{c}: {len(unique_test)}')

sex: 0
cabin: 37
CabinReduced: 1


Rozklad jest rownomierny dla cechy sex.

<h1> Zadanie 6 </h1>
Tworzenie slownikow, ktore wykorzystamy do kodowania:

In [34]:
Dict1 = {j: i for i, j in enumerate(df.sex.unique())}
Dict2 = {j: i for i, j in enumerate(df.cabin.unique())}
Dict3 = {j: i for i, j in enumerate(df.CabinReduced.unique())}

Sprawdzenie zawartosci utworzonego slownika.

In [35]:
import itertools

dict(itertools.islice(Dict3.items(), 10))

{'B': 0, 'C': 1, 'E': 2, 'D': 3, 'A': 4, nan: 5, 'T': 6, 'F': 7, 'G': 8}

Mozemy to zrobic w nieco ladniejszy sposob za pomoca funkcji.

In [36]:
from typing import Dict

def Encode(Train_set: pd.Series, Test_set: pd.Series) -> Dict:
    if Train_set.empty or Test_set.empty:
        raise ValueError("Empty set passed!")

    TrainTest = pd.concat([Train_set, Test_set])
    return {j: i for i, j in enumerate(TrainTest.unique())}

Sprawdzmy, czy funkcja dziala poprawnie.

In [37]:
dict_test = Encode(X_train['CabinReduced'], X_test['CabinReduced'])
dict(itertools.islice(dict_test.items(), 10))

{nan: 0, 'E': 1, 'C': 2, 'D': 3, 'B': 4, 'A': 5, 'F': 6, 'T': 7, 'G': 8}

<h1> Zadanie 7 </h1>
Tworzymy liste slownikow dla calego zbioru X.

In [38]:
dictionaries_list = [Encode(X_train[name], X_test[name]) for name in col_name]

Mapujemy, najpierw dla X_train.

In [39]:
X_train_mapped=[X_train[col].map(dictionaries_list[i]) for i, col in enumerate(X_train.columns)]
X_train_mapped[0]

501     0
588     0
402     0
1193    1
686     0
       ..
763     0
835     1
1216    0
559     0
684     0
Name: sex, Length: 916, dtype: int64

Dla ladniejszego wyswietlania mozemy zrobic:

In [40]:
X_train_mapped = pd.concat(X_train_mapped, axis = 1)
X_train_mapped

sex  cabin  CabinReduced
501     0      0             0
588     0      0             0
402     0      0             0
1193    1      0             0
686     0      0             0
...   ...    ...           ...
763     0      0             0
835     1      0             0
1216    0      0             0
559     0      0             0
684     0      0             0

[916 rows x 3 columns]

Analogicznie postepujemy dla X_test.

In [41]:
X_test_mapped=[X_test[col].map(dictionaries_list[i]) for i, col in enumerate(X_test.columns)]
X_test_mapped=pd.concat(X_test_mapped, axis=1)
X_test_mapped

sex  cabin  CabinReduced
1139    1      0             0
533     0      0             0
459     1      0             0
1150    1      0             0
393     1      0             0
...   ...    ...           ...
914     1      0             0
580     0      0             0
1080    1      0             0
1249    1    186             6
88      0      0             0

[393 rows x 3 columns]

<h1> Zadanie 8 </h1>

Ilosci brakujacych wartosci w zmodyfikowanych zbiorach.

In [42]:
print(f"Braki w X_train po mapowaniu: \n{X_train_mapped.isna().sum()}\n")
print(f"Braki w X_test po mapowaniu: \n{X_test_mapped.isna().sum()}")

Braki w X_train po mapowaniu: 
sex             0
cabin           0
CabinReduced    0
dtype: int64

Braki w X_test po mapowaniu: 
sex             0
cabin           0
CabinReduced    0
dtype: int64


W naszym przypadku braki nie wystepuja - NaN zostaly juz wczesniej zamienione wartoscia liczbowa.

<h1> Zadanie 9 </h1>

W naszym przypadku braki danych nie wystapily, ale moglismy je latwo i szybko uzupelnic uzywajac np fillna(value). Dla obecnego przypadku zastapienie NaN'ow zerami moze zle wplynac na nasz model ML - mozemy zbytnio uogolnic informacje chocby o kabinach (duzo wartosci i brakow, uzupelniajac 0 wszystkie braki po kolei dojdzie do sytuacji, w ktorej bardzo duza czesc pasazerow podrozuje w jednej kabinie co nie jest prawda - srednio 7 osob na kabine). Czasami natomiast moze byc to dobre rozwiazanie, czasami lepiej bedzie obliczyc srednia z sasiednich wartosci. Wszystko zalezy od danych, z jakimi mamy do czynienia.

<h1> Zadanie 10 </h1>

Porownanie unikalnych wartosci w zbiorze treningowym i testowym.

In [43]:
train_test = [X_train, X_test]
a = [[len(test_or_train[name].unique()) for name in col_name] for test_or_train in train_test]

train_test_names = ['X_train', 'X_test']
for i in range(2):
    print(f'{train_test_names[i]}:')
    for j in range(3):
        print(f'{col_name[j]}: {a[i][j]}')
    print('--------------------------------')

X_train:
sex: 2
cabin: 151
CabinReduced: 9
--------------------------------
X_test:
sex: 2
cabin: 71
CabinReduced: 8
--------------------------------


In [44]:
train_test_mapped = [X_train_mapped, X_test_mapped]
b = [[len(test_or_train[name].unique()) for name in col_name] for test_or_train in train_test_mapped]

train_test_names = ['X_train_mapped', 'X_test_mapped']
for i in range(2):
    print(f'{train_test_names[i]}:')
    for j in range(3):
        print(f'{col_name[j]}: {a[i][j]}')
    print('--------------------------------')

X_train_mapped:
sex: 2
cabin: 151
CabinReduced: 9
--------------------------------
X_test_mapped:
sex: 2
cabin: 71
CabinReduced: 8
--------------------------------


Wykonany proces ma ogromny wplyw na jakosc modelu i wynik predykcji - zbyt duza liczba etykiet prowadzi do overfittingu, dodatkowo chroni przed dominacja jednych zmiennych nad drugimi.

In [45]:
X_train = pd.concat([X_train, X_train_mapped], axis = 1)
X_train.columns = ['sex', 'cabin', 'CabinReduced', 'sex_mapped', 'cabin_mapped', 'CabinReduced_mapped']
X_test = pd.concat([X_test, X_test_mapped], axis = 1)
X_test.columns = ['sex', 'cabin', 'CabinReduced', 'sex_mapped', 'cabin_mapped', 'CabinReduced_mapped']

In [46]:
X_train.to_csv('X_train.csv')
X_test.to_csv('X_test.csv')
y_train.to_csv('y_train.csv')
y_test.to_csv('y_test.csv')